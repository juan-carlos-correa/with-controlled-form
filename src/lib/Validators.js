export const Validators = (value, errors = []) => ({
  isRequired: function (verify) {
    if (verify && !value) {
      errors.push('Este valor es requerido');
    }

    return Validators(value, errors);
  },

  isEmail: function () {
    const reg = /\S+@\S+\.\S+/;

    if (!reg.test(value)) {
      errors.push('Este valor debe ser un email del tipo name@mail.com');
    }

    return Validators(value, errors);
  },

  isMinLength: function (size) {
    if (value.length < size) {
      errors.push(`Este valor debe contener al menos ${size} caracteres`);
    }

    return Validators(value, errors);
  },

  isMaxLength: function (size) {
    if (value.length > size) {
      errors.push(`Este valor debe contener m√°ximo ${size} caracteres`);
    }

    return Validators(value, errors);
  },

  isEqual: function (str) {
    if (value !== str) {
      errors.push('Los valores son diferentes');
    }

    return Validators(value, errors);
  },

  validate: function (globalValues, valueToCheck, rules = {}) {
    let validator = Validators(valueToCheck);

    const isRequired = rules.isRequired;

    for (let rule in rules) {
      let valueToValidate = rules[rule];

      if (!isRequired && rule === 'isRequired') {
        continue;
      }

      if (!isRequired && !valueToCheck) {
        continue;
      }

      if (rule === 'isEqual') {
        valueToValidate = globalValues[valueToValidate];
      }

      validator = validator[rule](valueToValidate);
    }

    return validator.getResult();
  },

  getResult: function () {
    const isValid = errors.length === 0;

    return { isValid, errors };
  }
})

export default Validators;
