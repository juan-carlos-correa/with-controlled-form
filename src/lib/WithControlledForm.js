import React, { Component } from 'react';
import { Validators } from './Validators';

export const WithControlledForm = (FormComponent, state = {}, formValidations = {}) => (
  class WithFormMethodsHOC extends Component {
    constructor (props) {
      super(props);

      this.state = {
        values: { ...state },
        errors: this._stateToErrors(state),
      };
    }

    _stateToErrors = (state) => {
      const errors = {};

      const keys = Object.keys(state);
      for (let i = 0; i < keys.length; i += 1) {
        errors[keys[i]] = [];
      }

      return errors;
    }

    _validateForm = (values) => {
      const result = Validators.validate(values, formValidations);
      this.setState({ errors: result.errors });
      return result.isValid;
    }

    isFormClean = () => {
      const { values } = this.state;
      const isValue = Object.values(values).filter(val => val.length > 0);
      return isValue.length === 0;
    }

    cleanForm = () => {
      const newValues = {};
      const { values } = this.state;

      for (let val in values) {
        newValues[val] = '';
      }

      this.setState({
        values: newValues,
        errors: this._stateToErrors(newValues),
      });
    }

    handleChange = (e) => {
      const { name, value } = e.target;
      const { values } = this.state;

      values[name] = value;
      this.setState({ values });
    }

    handleCheckboxChange = (e) => {
      const { name, checked } = e.target;
      const { values } = this.state;

      values[name] = checked;
      this.setState({ values });
    }

    handleBlur = (e) => {
      e.preventDefault();

      const { errors } = this.state;
      const { name, value } = e.target;
      const validations = formValidations[name];
      const result = Validators.validateOne(value, validations);
      errors[name] = !result.isValid ? result.errors[0] : [];
      this.setState({ errors });
    }

    handleSubmit = (e) => {
      e.preventDefault();
      const { values } = this.state;
      const { handleSubmit } = this.props;

      if (this._validateForm(values)) {
        handleSubmit(values);
      }
    }

    render () {
      return (
        <FormComponent
          {...this.props}
          {...this.state}
          cleanForm={this.cleanForm}
          isFormClean={this.isFormClean}
          handleChange={this.handleChange}
          handleCheckboxChange={this.handleCheckboxChange}
          handleBlur={this.handleBlur}
          handleSubmit={this.handleSubmit}
        />
      )
    }
  }
);
