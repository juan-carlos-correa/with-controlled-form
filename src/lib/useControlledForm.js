import React, { useState } from 'react'
import { Validators } from '../lib/Validators'

const _stateToErrors = (state) => {
  const errors = {};
  const names = Object.keys(state);

  for (let i = 0; i < names.length; i += 1) {
    const keyName = names[i];
    errors[keyName] = [];
  }

  return errors;
}

export const useControlledForm = (initialState, formValidations) => {
  const [values, setValues] = useState({ ...initialState });
  const [errors, setErrors] = useState(_stateToErrors(initialState));

  const handleChange = ({ target }) => {
    const { name, value } = target;

    setValues({ ...values, [name]: value });
  }

  const handleSubmit = (_handleSubmit) => {
    if (_validateForm(values)) {
      _handleSubmit(values);
    }
  }

  const _validateForm = (valuesToValidate) => {
    const result = Validators.validate(valuesToValidate, formValidations);
    setErrors({ errors: result.errors });
    return result.isValid;
  }

  return {values, errors, handleChange, handleSubmit };
}
