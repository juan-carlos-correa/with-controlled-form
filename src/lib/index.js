import WithControlledForm from './WithControlledForm';
export default WithControlledForm;
