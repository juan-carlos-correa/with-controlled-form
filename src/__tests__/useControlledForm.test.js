import React from 'react';
import { mount } from 'enzyme';
import Form from '../demo/components/Form';
import { useControlledForm } from '../lib';

const formState = {
  email: '',
  phoneNumber: '',
};

const formValidations = {
  email: {
    isRequired: true,
    isMinLength: 3,
    isMaxLength: 30,
    isEmail: [true, 'email is not valid!'],
  },
  phoneNumber: {
    isRequired: true,
    isMinLength: 6,
    isMaxLength: 15,
  }
};

const FormWithUseControlledFormHook = ({ myHandleSubmit }) => {
  const {
    values,
    errors,
    handleChange,
    handleSubmit
   } = useControlledForm(formState, formValidations);

  return <Form
    values={values}
    errors={errors}
    handleChange={handleChange}
    handleSubmit={() => handleSubmit(myHandleSubmit)}
  />
}

describe('useControlledForm hook', () => {
  describe('valid render and correct submit', () => {
    test('render without crash', () => {
      const handleSubmitMock = jest.fn();
      const wrapper = mount(<FormWithUseControlledFormHook myHandleSubmit={handleSubmitMock} />);

      expect(wrapper).toMatchSnapshot();

      wrapper.unmount();
    });

    test('call submit handler function properly with valid values', () => {
      const handleSubmitMock = jest.fn();
      const wrapper = mount(<FormWithUseControlledFormHook myHandleSubmit={handleSubmitMock} />);

      const email = wrapper.find('#email');
      expect(email).toHaveLength(1);

      email.simulate('change', { target: { value: 'foo@mail.com', name: 'email' } });

      const phoneNumber = wrapper.find('#phoneNumber');
      expect(phoneNumber).toHaveLength(1);

      phoneNumber.simulate('change', { target: { value: '5512345678', name: 'phoneNumber' } });

      const form = wrapper.find('form');
      form.simulate('submit');

      expect(handleSubmitMock).toHaveBeenCalledTimes(1);

      wrapper.unmount();
    });
  })

  describe('input validations on submit', () => {
    test('required fields', () => {
      const handleSubmitMock = jest.fn();
      const wrapper = mount(<FormWithUseControlledFormHook myHandleSubmit={handleSubmitMock} />);

      const form = wrapper.find('form');
      form.simulate('submit');

      const emailErrorMessage = wrapper.find('#email-error-message');
      expect(emailErrorMessage.text()).toContain('This value is required');

      const phoneNumberErrorMessage = wrapper.find('#phoneNumber-error-message');
      expect(phoneNumberErrorMessage.text()).toContain('This value is required');

      expect(handleSubmitMock).toHaveBeenCalledTimes(0);

      wrapper.unmount();
    });

    test('isMinLength', () => {
      const handleSubmitMock = jest.fn();
      const wrapper = mount(<FormWithUseControlledFormHook myHandleSubmit={handleSubmitMock} />);

      const email = wrapper.find('#email');
      email.simulate('change', { target: { value: 'fo', name: 'email' } });

      const phoneNumber = wrapper.find('#phoneNumber');
      phoneNumber.simulate('change', { target: { value: '123', name: 'phoneNumber' } });

      const form = wrapper.find('form');
      form.simulate('submit');

      const emailErrorMessage = wrapper.find('#email-error-message');
      expect(emailErrorMessage.text()).toContain('This value must contain at least 3 characters');

      const phoneNumberErrorMessage = wrapper.find('#phoneNumber-error-message');
      expect(phoneNumberErrorMessage.text()).toContain('This value must contain at least 6 characters');

      expect(handleSubmitMock).toHaveBeenCalledTimes(0);

      wrapper.unmount();
    });

    test('isMaxLength', () => {
      const handleSubmitMock = jest.fn();
      const wrapper = mount(<FormWithUseControlledFormHook myHandleSubmit={handleSubmitMock} />);

      const email = wrapper.find('#email');
      email.simulate('change', { target: { value: 'foooooooooooooooooooooooo@mail.com', name: 'email' } });

      const phoneNumber = wrapper.find('#phoneNumber');
      phoneNumber.simulate('change', { target: { value: '1234567891234567', name: 'phoneNumber' } });

      const form = wrapper.find('form');
      form.simulate('submit');

      const emailErrorMessage = wrapper.find('#email-error-message');
      expect(emailErrorMessage.text()).toContain('This value must contain a maximum of 30 characters');

      const phoneNumberErrorMessage = wrapper.find('#phoneNumber-error-message');
      expect(phoneNumberErrorMessage.text()).toContain('This value must contain a maximum of 15 characters');

      expect(handleSubmitMock).toHaveBeenCalledTimes(0);

      wrapper.unmount();
    });

    test('isEmail with custom message', () => {
      const handleSubmitMock = jest.fn();
      const wrapper = mount(<FormWithUseControlledFormHook myHandleSubmit={handleSubmitMock} />);

      const email = wrapper.find('#email');
      email.simulate('change', { target: { value: 'foomail.com', name: 'email' } });

      const form = wrapper.find('form');
      form.simulate('submit');

      const emailErrorMessage = wrapper.find('#email-error-message');
      expect(emailErrorMessage.text()).toContain('email is not valid!');

      expect(handleSubmitMock).toHaveBeenCalledTimes(0);

      wrapper.unmount();
    });
  });
});
