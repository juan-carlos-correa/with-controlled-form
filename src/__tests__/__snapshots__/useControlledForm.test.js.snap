// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useControlledForm hook valid render and correct submit render without crash 1`] = `
<FormWithUseControlledFormHook
  myHandleSubmit={[MockFunction]}
>
  <Form
    errors={
      Object {
        "email": Array [],
        "phoneNumber": Array [],
      }
    }
    handleChange={[Function]}
    handleSubmit={[Function]}
    values={
      Object {
        "email": "",
        "phoneNumber": "",
      }
    }
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-element"
      >
        <label
          htmlFor="email"
        >
          Email:
        </label>
        <input
          className="input"
          id="email"
          name="email"
          onChange={[Function]}
          type="text"
          value=""
        />
        <div
          className="error-message"
          id="email-error-message"
        />
      </div>
      <div
        className="form-element"
      >
        <label
          htmlFor="phoneNumber"
        >
          Phone number:
        </label>
        <input
          className="input"
          id="phoneNumber"
          name="phoneNumber"
          onChange={[Function]}
          type="tel"
          value=""
        />
        <div
          className="error-message"
          id="phoneNumber-error-message"
        />
      </div>
      <div
        className="text-center"
      >
        <button
          className="button"
          id="send"
          type="submit"
        >
          Send
        </button>
      </div>
    </form>
  </Form>
</FormWithUseControlledFormHook>
`;
